### ISSUE TITLE
Ensure we can send JSON null in payloads

### ISSUE THREAD

**Issue by jhendrixMSFT** [Azure/autorest.go#432]
This is to support `PATCH` where the caller wants to clear some value by sending a `null`.

Latest idea is to remove the `,omitempty` from the JSON tag and use a custom marshaller along with some pre-defined sentinel values in `azcore`, e.g. `azcore.NullString`.  When the sentinel null is set in the payload the null is included, else any null field is omitted.

**Comment by jhendrixMSFT** [comment]
https://gist.github.com/jhendrixMSFT/2340616d2583e853f3096e826953babb

**Comment by tombuildsstuff** [comment]
@jhendrixMSFT out of interest, why not use Go's native `null` type? Is this because the Swagger doesn't contain the necessary information to be able to determine when this should be a pointer/empty value?

**Comment by jhendrixMSFT** [comment]
@tombuildsstuff you mean simply leaving the field as `nil` or something different?

**Comment by tombuildsstuff** [comment]
@jhendrixMSFT yes, sorry - nil not null.

At first glance it feels like having a custom type for something so foundational (to Go) could be a common source of confusion where there's a native way to achieve this (and this is a pointer)?

By contrast generating the models with a mix of fields a as pointers and others as primitive types (rather than everything as pointers) feels like a more native approach to this?

**Comment by jhendrixMSFT** [comment]
I agree this solution is sub-optimal.  In addition, it breaks down for user-defined types.  However I don't know that using pointer/non-pointers will fix all cases either.

For PATCH there are two cases.  Running with the Widget example, say I get back the following.

```json
{
  "name": "creeper",
  "color: "green",
  "value": 1
}
```

For PATCH rules  (and UPSERT I believe), I need to omit the fields I want to preserve from the payload and send a null for the ones I want to clear.  So if I simply want to clear `value` I send the following payload.

```json
{
  "value": null
}
```

Unfortunately using a mix of pointer/non-pointer types has some ambiguous cases.  If the `Value` field is a non-pointer, how do I indicate I want to clear it (sentinel values are ambiguous)?  And if `Name` and `Color` are non-pointers how do I indicate I want to preserve their values (empty strings are ambiguous)?

Another idea I had was to have a method on types used for PATCH/UPSERT that takes a bit field indicating which fields to send as null.  It feels clunky though.

Maybe setter methods on such types?

Totally open to other ideas as this has been a long-standing problem with our current SDK.

**Comment by jhendrixMSFT** [comment]
Presumably I can also leave any fields as-is to preserve their values.

**Comment by jhendrixMSFT** [comment]
I think this can be solved via reflection, if you set a field to pointer-to-zero-value we can detect this in our custom marshaller.

Latest gist: https://gist.github.com/jhendrixMSFT/2340616d2583e853f3096e826953babb

**Comment by jhendrixMSFT** [comment]
Scratch that, doesn't work for meaningful zero-values.

**Comment by jhendrixMSFT** [comment]
Fixed in upcoming 4.0.0-preview.15
